### PR TITLE
test(pacquet): coverage

### PR DESCRIPTION
**Pull request by KSXGitHub**
Adds 44 unit + integration tests to close the easier targets from a `cargo llvm-cov` pass over the `pacquet/` workspace, lifting line coverage on the touched files without refactoring any production code.

## What's covered now

- **Pure logic** — `lockfile::resolved_dependency` (`From` / `Display` / `Serialize` / `TryFrom<Cow>` arms), `lockfile::{freshness, pkg_name, snapshot_dep_ref}`, `registry::package` (`PartialEq` / `latest` / `pinned_version`), `graph-hasher::{engine_name, object_hasher}` (Node-version detection + null / bool / array dispatch), `package-manifest` (engines-runtime conversion edge cases), `fs::file_mode`, `executor::execute_shell`.
- **I/O error classifications via real fixtures** — `lockfile::save_lockfile`, `workspace-state`, `patching::apply`, and `package-manifest::safe_read_package_json_from_dir` get coverage for their typed `CreateDir` / `RemoveFile` / `RenameFile` / `ReadFile` / `ParseJson` arms by planting a regular file or directory where the writer expects something else. No DI fakes needed — these branches are reachable through real filesystem state, which matches the "prefer real fixtures" rule the new style guide added in #11708.
- **New integration tests** — `pacquet run <script>` (happy path, extra-arg forwarding, missing-script with and without `--if-present`).

## Out of scope

The bigger gaps that *do* need the `Sys: <FsCapability + …>` seam from the [DI section of `CODE_STYLE_GUIDE.md`](https://github.com/pnpm/pnpm/blob/02f8138f/pacquet/CODE_STYLE_GUIDE.md#dependency-injection-for-tests) — `fs::ensure_file` rename retry / exhausted temp paths, `install_package_by_snapshot.rs` binary / git / directory resolution, `cmd-shim::link_bins` remaining error arms — are left for follow-ups. They need new `Fs*` capabilities on `Host` or fetcher fakes, which is more invasive than a tests-only PR should be.

## Drive-by

- Bumped `hex_decode` in `graph-hasher::tests` to `is_multiple_of` so test-time clippy stays clean on Rust 1.95.
- Dropped a trailing comma in a single-line `assert!` in `package-manifest/src/tests.rs` to satisfy Dylint's `perfectionist::macro-trailing-comma`.
- Merged `main` to pick up #11708.

## Test plan

- [x] `cargo nextest run` on every touched crate after merging `main` — 321 tests pass, including all the new ones.
- [x] `cargo clippy --locked --tests` on touched crates passes with `--deny warnings`.
- [x] `cargo fmt --check` clean.

---
Written by an agent (Claude Code, claude-opus-4-7).

## Summary by CodeRabbit

* **Tests**
  * Added many unit and integration tests across the project: CLI script execution and argument forwarding, Unix-gated shell/engine detection checks, file-mode and filesystem edge cases, hash/serialization/manifest/lockfile behaviors, patch-application error cases, package/version selection, and workspace state error handling.

* **Chores**
  * Added a development dependency for test utilities.

[Review Change Stack](https://app.coderabbit.ai/change-stack/pnpm/pnpm/pull/11710?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: pacquet/crates/cli/tests/run.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    let manifest_path = workspace.join("package.json");
+    let marker_path = workspace.join("args.txt");
+    // `printf %s "$1"` writes the first argument into the marker,
+    // letting the assertion below pin the exact argument flow.
+    let manifest = format!(
+        r#"{{
+            "name": "test",
+            "version": "0.0.0",
+            "scripts": {{ "echo-args": "sh -c 'printf %s \"$1\" > {marker}' --" }}
+        }}"#,
```
Comment (KSXGitHub) [inline pacquet/crates/cli/tests/run.rs]: Claude Code,

Do you think `json!({ ... }).to_string()` would have work better here?

Or do you think this test should just import the `Manifest` type from somewhere?

File: pacquet/crates/cli/tests/run.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+/// fails. Mirrors `pnpm run` behavior — the missing-script error
+/// comes from `PackageManifest::script`, surfaced through the
+/// `RunArgs::run` `wrap_err` chain.
+#[test]
+fn run_errors_on_missing_script_without_if_present() {
+    let CommandTempCwd { pacquet, root, workspace, .. } = CommandTempCwd::init();
+    let manifest_path = workspace.join("package.json");
+    fs::write(
+        &manifest_path,
+        r#"{"name": "test", "version": "0.0.0", "scripts": {"build": "echo built"}}"#,
```
Comment (KSXGitHub) [inline pacquet/crates/cli/tests/run.rs]: Same question as <https://github.com/pnpm/pnpm/pull/11710/changes#r3256855819>.

File: pacquet/crates/cli/tests/run.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+/// and pacquet exits cleanly. Required for orchestration tools
+/// that probe optional scripts without wanting to fail the
+/// pipeline.
+#[test]
+fn run_with_if_present_is_a_noop_for_missing_script() {
+    let CommandTempCwd { pacquet, root, workspace, .. } = CommandTempCwd::init();
+    let manifest_path = workspace.join("package.json");
+    fs::write(
+        &manifest_path,
+        r#"{"name": "test", "version": "0.0.0", "scripts": {"build": "echo built"}}"#,
```
Comment (KSXGitHub) [inline pacquet/crates/cli/tests/run.rs]: Same question as <https://github.com/pnpm/pnpm/pull/11710/changes#r3256855819>.

File: pacquet/crates/patching/src/apply/tests.rs [hunk 4]
```diff
@@ -338,3 +338,122 @@ deleted file mode 100644
...
+    let patch = write_patch(
+        patch_dir.path(),
+        "\
+diff --git a/partial.txt b/partial.txt
+deleted file mode 100644
+--- a/partial.txt
++++ /dev/null
+@@ -1 +0,0 @@
+-going away
+",
```
Comment (KSXGitHub) [inline pacquet/crates/patching/src/apply/tests.rs]: I prefer `text-block-macros`.

File: pacquet/crates/patching/src/apply/tests.rs [hunk 5]
```diff
@@ -338,3 +338,122 @@ deleted file mode 100644
...
+    fs::write(patched.path().join("from.txt"), "hello\n").unwrap();
+    let patch_dir = tempdir().unwrap();
+    let patch = write_patch(
+        patch_dir.path(),
+        "\
+diff --git a/from.txt b/to.txt
+similarity index 100%
+rename from from.txt
+rename to to.txt
+",
```
Comment (KSXGitHub) [inline pacquet/crates/patching/src/apply/tests.rs]: Same as <https://github.com/pnpm/pnpm/pull/11710/changes#r3256861076>.

File: pacquet/crates/patching/src/apply/tests.rs [hunk 6]
```diff
@@ -338,3 +338,122 @@ deleted file mode 100644
...
+    let patch = write_patch(
+        patch_dir.path(),
+        "\
+diff --git a/blocker/nested.txt b/blocker/nested.txt
+new file mode 100644
+--- /dev/null
++++ b/blocker/nested.txt
+@@ -0,0 +1 @@
++hi
+",
```
Comment (KSXGitHub) [inline pacquet/crates/patching/src/apply/tests.rs]: Same as <https://github.com/pnpm/pnpm/pull/11710/changes#r3256861076>.

File: pacquet/crates/cli/tests/run.rs (hunk 1 above)
Comment (KSXGitHub) [reply pacquet/crates/cli/tests/run.rs]: Switched all three `package.json` fixtures to `json!({...}).to_string()` in cca10a6 — keeps the shape inline and lets serde handle path escaping. Importing the real `PackageManifest` type would force me to populate every required field just to write a JSON file with three keys, so the lighter `json!` macro felt like a better fit here.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: pacquet/crates/cli/tests/run.rs (hunk 2 above)
Comment (KSXGitHub) [reply pacquet/crates/cli/tests/run.rs]: Resolved in cca10a6 — same as the linked thread.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: pacquet/crates/cli/tests/run.rs (hunk 3 above)
Comment (KSXGitHub) [reply pacquet/crates/cli/tests/run.rs]: Resolved in cca10a6 — same as the linked thread.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: pacquet/crates/patching/src/apply/tests.rs (hunk 4 above)
Comment (KSXGitHub) [reply pacquet/crates/patching/src/apply/tests.rs]: Converted all three new patch fixtures to `text_block_fnl!` in cca10a6 (added `text-block-macros` to the crate's dev-deps). Used the `_fnl` variant because git-diff parsers need the trailing newline. Left the pre-existing `IS_POSITIVE_PATCH` and the other older fixtures in this file untouched to keep the diff scoped to the lines flagged.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: pacquet/crates/patching/src/apply/tests.rs (hunk 5 above)
Comment (KSXGitHub) [reply pacquet/crates/patching/src/apply/tests.rs]: Resolved in cca10a6 — same as the linked thread.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: pacquet/crates/patching/src/apply/tests.rs (hunk 6 above)
Comment (KSXGitHub) [reply pacquet/crates/patching/src/apply/tests.rs]: Resolved in cca10a6 — same as the linked thread.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: pacquet/crates/executor/src/lib.rs [hunk 7]
```diff
@@ -34,3 +34,27 @@ pub fn execute_shell(command: &str) -> Result<(), ExecutorError> {
 
     Ok(())
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (zkochan) [inline pacquet/crates/executor/src/lib.rs]: wasn't the convention to always put tests in a separate file?

Comment (KSXGitHub) [reply pacquet/crates/executor/src/lib.rs]: Right — fixed in 50c39ca. Moved the `executor` and `fs::file_mode` tests into their conventional `src/<parent>/tests.rs` locations as documented in `CODE_STYLE_GUIDE.md` ("Unit test file layout"). My mistake on the inline `mod tests { ... }` block.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: pacquet/crates/executor/src/lib.rs [hunk 8]
```diff
@@ -34,3 +34,27 @@ pub fn execute_shell(command: &str) -> Result<(), ExecutorError> {
...
+    }
+
+    /// A non-zero exit from the spawned command still surfaces as
+    /// `Ok(())`: `execute_shell` only fails on I/O errors from
+    /// spawn or wait. Pnpm's `run` semantics treat the script's
+    /// exit code as its own concern, not a wrapper failure.
+    #[cfg(unix)]
+    #[test]
+    fn execute_shell_propagates_nonzero_exit_as_ok() {
+        execute_shell("exit 1").expect("wait succeeds even on nonzero exit");
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply pacquet/crates/executor/src/lib.rs]: Good catch. Dropped this test in 50c39ca. The "non-zero exit → `Ok`" behavior in `execute_shell` is indeed a bug relative to pnpm/npm's `run` semantics (which propagate the script's status as the CLI's exit code), and pinning it as a regression test would block the eventual fix. The fix itself belongs in its own PR — outside the scope of a tests-only change.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: pacquet/crates/graph-hasher/src/engine_name.rs [hunk 9]
```diff
@@ -236,4 +236,28 @@ mod tests {
...
         assert!(parts[2][4..].parse::<u32>().is_ok(), "node version must be numeric: {name:?}");
     }
+
+    /// `detect_node_version` returns the full version string with
+    /// the leading `v` stripped. `node` is a hard prerequisite for
+    /// the test suite — if it isn't on `PATH` that's a test-env
+    /// bug, so we `expect` rather than skip.
+    #[test]
+    fn detect_node_version_strips_leading_v() {
+        let version = detect_node_version().expect("`node` must be on PATH for the test suite");
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply pacquet/crates/graph-hasher/src/engine_name.rs:246]: Intentionally left as-is. A human reviewer (KSXGitHub) explicitly asked in [r3255978...] for these tests to fail loudly when `node` is missing rather than skip silently, on grounds that `node` is a documented prerequisite of the test suite (see `pacquet/CONTRIBUTING.md` "Setup → Prerequisites"). That choice landed as 5d0987c. Reverting to a tolerant skip would undo it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: pacquet/crates/cli/tests/run.rs [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+#[cfg(unix)]
+#[test]
+fn run_executes_declared_script() {
+    let CommandTempCwd { pacquet, root, workspace, .. } = CommandTempCwd::init();
+    let manifest_path = workspace.join("package.json");
+    let marker_path = workspace.join("marker.txt");
+    let manifest = json!({
+        "name": "test",
+        "version": "0.0.0",
+        "scripts": { "touch-marker": format!("touch {}", marker_path.display()) },
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply pacquet/crates/cli/tests/run.rs]: Fixed in 50c39ca. Wrapped the temp-path interpolation in `"..."` in both `cli/tests/run.rs` shell fixtures so a path with a space (the common macOS `/var/folders/...` shape) doesn't split the redirect/`touch` argument. Stopped short of a full shell-escape because `tempfile::tempdir()` paths can't contain `'` or `"` in practice — quoting handles the realistic case.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: pacquet/crates/executor/src/lib.rs [hunk 11]
```diff
@@ -34,3 +34,27 @@ pub fn execute_shell(command: &str) -> Result<(), ExecutorError> {
 
     Ok(())
 }
+
+#[cfg(test)]
+mod tests {
+    use super::execute_shell;
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply pacquet/crates/executor/src/lib.rs]: Fixed in 50c39ca by gating `mod tests` itself on `#[cfg(all(test, unix))]` at the parent's declaration site. Since both tests in the module spawn `sh` and are Unix-only, the cleanest fix is to compile the whole module out on Windows rather than to scatter `#[cfg(unix)]` across each test.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: pacquet/crates/graph-hasher/src/engine_name.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply pacquet/crates/graph-hasher/src/engine_name.rs:246]: Yes. Tolerant tests are kinda pointless.

I wonder. Amongst the existing AI-generated tests, how many are mistakenly "tolerant" like the one Copilot suggested.

**Review by zkochan — APPROVED** [review]
